transform/startf/transform.go

```diff
@@ -168,229 +144,194 @@ func ExecScript(ctx context.Context, target *dataset.Dataset, opts ...func(o *Ex
+	return func(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+		var refstr starlark.String
+		if err := starlark.UnpackArgs("load_version", args, kwargs, "ref", &refstr); err != nil {
+			return starlark.None, err
 		}
-		defined[name] = fn
-	}
-
-	return
-}
 
-func callDownloadFunc(t *transform, thread *starlark.Thread, ctx *skyctx.Context) (result starlark.Value, err error) {
-	httpGuard.EnableNtwk()
-	defer httpGuard.DisableNtwk()
-
-	var download *starlark.Function
-	if download, err = t.globalFunc("download"); err != nil {
-		if err == ErrNotDefined {
-			return starlark.None, nil
+		if r.dsLoader == nil {
+			return nil, fmt.Errorf("load_version function is not enabled")
 		}
-		return starlark.None, err
-	}
-
-	return starlark.Call(thread, download, starlark.Tuple{ctx.Struct()}, nil)
-}
 
-func callTransformFunc(t *transform, thread *starlark.Thread, ctx *skyctx.Context) error {
-	transform, err := t.globalFunc("transform")
-	if err != nil {
-		if err == ErrNotDefined {
-			return nil
+		ds, err := r.dsLoader.LoadDataset(ctx, refstr.GoString())
+		if err != nil {
+			return starlark.None, err
 		}
-		return err
-	}
-
-	d := skyds.NewDataset(t.target)
-	if _, err = starlark.Call(thread, transform, starlark.Tuple{d, ctx.Struct()}, nil); err != nil {
-		return err
-	}
 
-	// Which components were changed
-	if t.changeSet != nil {
-		changes := d.Changes()
-		for comp := range changes {
-			t.changeSet[comp] = changes[comp]
+		if target.Transform.Resources == nil {
+			target.Transform.Resources = map[string]*dataset.TransformResource{}
 		}
-	}
 
-	// assign the resulting dataframe to the dataset body
-	return d.AssignBodyFromDataframe()
-}
+		target.Transform.Resources[ds.Path] = &dataset.TransformResource{
+			// TODO(b5) - this should be a method on dataset.Dataset
+			// we should add an ID field to dataset, set that to the InitID, and
+			// add fields to dataset.TransformResource that effectively make it the
+			// same data structure as dsref.Ref
+			Path: fmt.Sprintf("%s/%s@%s", ds.Peername, ds.Name, ds.Path),
+		}
 
-func (t *transform) locals() starlark.StringDict {
-	return starlark.StringDict{
-		"load_dataset": starlark.NewBuiltin("load_dataset", t.LoadDataset),
-		"print":        starlark.NewBuiltin("print", t.print),
+		return stards.NewDataset(ds), nil
 	}
 }
 
-// ModuleLoader sums all loading assets to resolve a module name during transform execution
-func (t *transform) ModuleLoader(thread *starlark.Thread, module string) (dict starlark.StringDict, err error) {
-	if module == skyqri.ModuleName && t.skyqri != nil {
-		return t.skyqri.Namespace(), nil
+func (r *StepRunner) print(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var message starlark.String
+	if err := starlark.UnpackArgs("print", args, kwargs, "message", &message); err != nil {
```

Comment:
Something that the built-in starlark `print` function does is support for auto stringification. That means as long as a type implements `starlark.Stringer` it can be passed as the first argument to `print` and will be converted to a string automatically. As written, this function instead requires this:

```
df = dataframe.DataFrame()
print('%s' % df)
```

rather than just:

```
df = dataframe.DataFrame()
print(df)
```

Would be nice if we could add this functionality as well. Doesn't have to happen in this refactoring PR, but a TODO would be nice.

Reply:
ugh good point. our print is terrible. fixed.